server-cloud/server/auth.go

```diff
@@ -0,0 +1,195 @@
+
+	sess.Values["state"] = state
+
+	if err := sess.Save(c.Request(), c.Response()); err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+	}
+
+	return c.Redirect(http.StatusTemporaryRedirect, s.auth.AuthCodeURL(state))
+}
+
+func (s *Server) callback(c echo.Context) error {
+	sess, err := session.Get(authSessionName, c)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+	}
+
+	if c.QueryParam("state") != sess.Values["state"] {
+		return c.String(http.StatusBadRequest, "Invalid state parameter.")
+	}
+
+	// Exchange an authorization code for a token.
+	token, err := s.auth.Exchange(c.Request().Context(), c.QueryParam("code"))
+	if err != nil {
+		return c.String(http.StatusUnauthorized, "Failed to convert an authorization code into a token.")
+	}
+
+	idToken, err := s.auth.VerifyIDToken(c.Request().Context(), token)
+	if err != nil {
+		return c.String(http.StatusInternalServerError, "Failed to verify ID Token.")
+	}
+
+	var profile map[string]interface{}
+	if err := idToken.Claims(&profile); err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	sess.Values["access_token"] = token.AccessToken
+	sess.Values["profile"] = profile
+
+	if err := sess.Save(c.Request(), c.Response()); err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	// Redirect to logged in page.
+	return c.Redirect(http.StatusTemporaryRedirect, "/auth/user")
+}
+
+func (s *Server) logoutCallback(c echo.Context) error {
+	sess, err := session.Get(authSessionName, c)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+	}
+
+	sess.Values["access_token"] = nil
+	sess.Values["profile"] = nil
+
+	if err := sess.Save(c.Request(), c.Response()); err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	return c.String(http.StatusOK, "user logged out")
+}
+
+func (s *Server) logout(c echo.Context) error {
+	logoutUrl, err := url.Parse("https://" + s.conf.Auth0Domain + "/v2/logout")
+	if err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	scheme := "http"
+	if c.Request().TLS != nil {
+		scheme = "https"
+	}
+
+	returnTo, err := url.Parse(scheme + "://" + c.Request().Host + "/auth/logout/callback")
+	if err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	parameters := url.Values{}
+	parameters.Add("returnTo", returnTo.String())
+	parameters.Add("client_id", s.conf.Auth0ClientID)
+	logoutUrl.RawQuery = parameters.Encode()
+
+	return c.Redirect(http.StatusTemporaryRedirect, logoutUrl.String())
+}
+
+func (s *Server) user(c echo.Context) error {
+	sess, err := session.Get(authSessionName, c)
+	if err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	profile := sess.Values["profile"]
+	return c.JSON(http.StatusOK, profile)
+}
+
+func GetAuth(c context.Context, conf Config) (*Authenticator, error) {
```

Comment:
Call this function `newAuthenticator` instead, and place it right after the `type Authenticator struct {` struct. Make it lowercase since it's not used externally, only inside this package.

In go, the common structure is a) define a struct, b) define "new" functions for creating the struct, c) define the members functions of the struct.